### PR TITLE
Fixed up some initial code to do evals based off synthetic questions …

### PR DESCRIPTION
**Pull request by ivanleomk**
Added some code to generate evaluations based on synthetic questions

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 66c3ed1210aa21994e6ef3627eed8e779efbc66b.  | 
|--------|--------|

### Summary:
This PR introduces an evaluation feature to the RAG application, modifies synthetic question generation, updates models, refines the chunking process, and removes print statements from metrics calculation functions.

**Key points**:
- Added 'evaluate' command in `/rag_app/cli.py`.
- Implemented evaluation functionality in new file `/rag_app/evaluate.py`.
- Modified synthetic question generation in `/rag_app/generate_synthetic_question.py`.
- Updated models in `/rag_app/models.py` to include 'QuestionAnswerPair' and 'EvaluationDataItem'.
- Updated chunking process in `/rag_app/src/chunking.py` to use additional metadata fields.
- Removed print statements from metrics calculation functions in `/rag_app/src/metrics.py`.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: rag_app/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    async def query_table(query: EmbeddedEvaluationItem):
+        results = table.search(query.embedding).limit(10).to_pydantic(TextChunk)
+        return QueryResult(results=results, source=query)
+
+    coros = [query_table(query) for query in queries]
+    return await asyncio.gather(*coros)
+
+
+@app.command(help="Evaluate document retrieval")
+def from_json(
```
Comment (jxnl) [inline rag_app/evaluate.py]: why is the function call from_json?

File: rag_app/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    assert (
+        Path(input_file_path).suffix == ".json"
+    ), "The output file must have a .json extension."
+    assert Path(db_path).exists(), f"Database path {db_path} does not exist"
+
+    with open(input_file_path, "r") as file:
+        data = json.load(file)
+        evaluation_data = [EvaluationDataItem(**json.loads(item)) for item in data]
+
+    embedded_queries = run(embed_test_queries(evaluation_data))
```
Comment (jxnl) [inline rag_app/evaluate.py:114]: this should be unwarpped by `embed_test_queries` flattening it is a implementation detail that is being leaked/

File: rag_app/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    flattened_queries = [item for sublist in embedded_queries for item in sublist]
+
+    query_results = run(fetch_relevant_results(flattened_queries, db_path, table_name))
+
+    for result in query_results:
```
Comment (jxnl) [inline rag_app/evaluate.py]: shoul ddefine a scoring function.

think about writing the pseudo code first before the whole function

dataset = load()
generate_results = process(dataset)
scores = score(generate_results)

but better function names

File: test.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+[
```
Comment (jxnl) [inline test.json]: this is not json. this should be jsonlines

File: rag_app/evaluate.py (hunk 1 above)
Comment (ivanleomk) [reply rag_app/evaluate.py]: It reads in the queries from a .json file but now that I'm seeing your other comments, I'll change the generation to be a .jsonl file

File: rag_app/evaluate.py (hunk 3 above)
Comment (ivanleomk) [reply rag_app/evaluate.py]: Ended up defining a scorer function which takes in a query and returns a score for it in the form of a dict { metric: value }

**Comment by ivanleomk** [comment]
Final result is something like this 

```
                                     MRR@3     MRR@5    MRR@10    MRR@20   NDCG@3    NDCG@5   NDCG@10   NDCG@20
chunk_id                                                                                                       
de18acba58bcd5ec9995871f0f69005d  1.000000  1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000
61d640330dcfac2c0a786d9ee5522375  1.000000  1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000
dbbec66cbef84b1763a53134c8e782da  0.500000  0.500000  0.500000  0.500000  0.63093  0.630930  0.630930  0.630930
26d677b5e3cd7db5e1971408839d0b65  1.000000  1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000
83fe7dd856a75e38a461563e6c71b5ab  1.000000  1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000
df602c06b072a9ee919034b1903bad01  1.000000  1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000
```

**Comment by ivanleomk** [comment]
I suspect a reason why some of these values are so high is because of repeated words in the question and answer

```
{"question":"Why is Silicon Valley tolerant of failure according to the text?","answer":"Silicon Valley is tolerant of failure because people value learning from failures. They believe that if an individual learns from their failures, they are a good bet for success in the long run.","chunk":"This is one reason Silicon Valley is so tolerant of failure. People in Silicon Valley aren't blindly tolerant of failure. They'll only continue to bet on you if you're learning from your failures. But if you are, you are in fact a good bet: maybe your company didn't grow the way you wanted, but you yourself have, and that should yield results eventually.","chunk_id":"61d640330dcfac2c0a786d9ee5522375"}
```

we can see here that the phrase `the reason why sillicon valley is so tolerant of failure` is repeated multiple times

**Pull request merged**